### PR TITLE
ROOT I/O performance related updates

### PR DESCRIPTION
**Pull request by amadio**
This branch contains modifications to address some performance issues identifies in JIRA issue [ROOT-8871](https://sft.its.cern.ch/jira/browse/ROOT-8871).

The main changes are:
 * Make `TClass::LoadClassInfo()` private and avoid locking the interpreter unnecessarily
 * Avoid interpreter lock in `TClass::GetListOfBases()` and return existing list if already available
 * Reduce scope of interpreter locks in several places where the lock is taken before necessary
 * Add locks where unprotected use of `gInterpreter` is made
 * Improve code clarity and performance of `TTree::Fill()`
 * Some typo and formatting fixes to improve conformance to coding conventions

### CODE REVIEW

**Comment by dpiparo** [comment]
Great PR, a very nice job. The first thing, even without going into the details, I'd like to note is that many changes to formatting have been applied. I wonder if those are really necessary and do not add noise/dificulties by cherry-picking.

File: core/meta/inc/TClass.h [hunk 1]
```diff
@@ -173,7 +176,8 @@ friend class TProtoClass;
    TListOfDataMembers *fData;            //linked list for data members
 
    std::atomic<TListOfEnums*> fEnums;        //linked list for the enums
-   TListOfFunctionTemplates  *fFuncTemplate; //linked list for function templates [Not public until implemented as active list]
+   TListOfFunctionTemplates  *fFuncTemplate; //linked list for function templates
```
Comment (Axel-Naumann) [inline core/meta/inc/TClass.h]: Note that these changes change the TDataMember's comment. I.e. it's a functionality change.

**Comment by amadio** [comment]
Hi Danilo. Yes, indeed, there are quite a few formatting changes. I was reading the whole file, so I thought I might as well fix the more glaring issues with formatting. I can separate the commits that are pure formatting and just push them first, leaving only real changes here if you prefer. Cherry picking shouldn't be difficult, however, as I did try to make the commits more or less independent. They do not interfere with each other.

File: core/meta/inc/TClass.h (hunk 1 above)
Comment (amadio) [inline core/meta/inc/TClass.h]: Oops, sorry, I will revert this part then. Is it ok to just put the bracketed text on the next line, or does it need to be all in one line?

File: core/meta/src/TClass.cxx [hunk 2]
```diff
@@ -4358,29 +4358,33 @@ Int_t TClass::GetNmethods()
...
-   if (version == 0) {
+   if (version == 0)
       version = fClassVersion;
-   }
-   Int_t ninfos = fStreamerInfo->GetSize();
-   if ((version < -1) || (version >= ninfos)) {
+
+   // If the StreamerInfo is assigned to the fLastReadInfo, we are
+   // guaranteed it was built and compiled.
+   if (sinfo && sinfo->GetClassVersion() == version)
```
Comment (Axel-Naumann) [inline core/meta/src/TClass.cxx]: That's a change in behavior (`version` might have been set to `fClassVersion` after your change but not before); could you document it in the commit log if it was intentional?

Comment (amadio) [inline core/meta/src/TClass.cxx]: This is intentional. When the requested version is 0, it really means the current version, so this let's the detection work properly and return the version already available (that's the point of this check). However, before the check would always fail, because fClassVersion is never 0 (at least that's how I interpreted it, correct me if I'm wrong).

Comment (Axel-Naumann) [inline core/meta/src/TClass.cxx]: > (which supposedly isn't itself equal to 0)
...
> (at least that's how I interpreted it, correct me if I'm wrong)

These are assumptions; they suggest that we are probably better off waiting for Philippe with these changes?

Comment (amadio) [inline core/meta/src/TClass.cxx]: Yes, definitely, I will not merge this before Philippe is back. I just want to give others a chance to review immediately to address any problems found before he comes back.

File: core/meta/inc/TClass.h (hunk 1 above)
Comment (amadio) [inline core/meta/inc/TClass.h]: Reverted.

File: core/meta/inc/TClass.h [hunk 3]
```diff
@@ -232,9 +235,9 @@ friend class TProtoClass;
 
    typedef void (*StreamerImpl_t)(const TClass* pThis, void *obj, TBuffer &b, const TClass *onfile_class);
 #ifdef R__NO_ATOMIC_FUNCTION_POINTER
-   mutable StreamerImpl_t fStreamerImpl;  //! Pointer to the function implementing the right streaming behavior for the class represented by this object.
+   mutable StreamerImpl_t fStreamerImpl; //! Pointer to the function implementing streaming for this class
```
Comment (pcanal) [inline core/meta/inc/TClass.h]: The original comment is intentionally pendantic 'this class' could be meaning 'the class named TClass' or the "the class represented by the instance of TClass"

File: core/meta/inc/TClass.h [hunk 4]
```diff
@@ -309,12 +311,14 @@ friend class TProtoClass;
...
 
+   void LoadClassInfo() const;
+
 protected:
-   TVirtualStreamerInfo     *FindStreamerInfo(TObjArray* arr, UInt_t checksum) const;
-   void                      GetMissingDictionariesForBaseClasses(TCollection& result, TCollection& visited, bool recurse);
-   void                      GetMissingDictionariesForMembers(TCollection& result, TCollection& visited, bool recurse);
-   void                      GetMissingDictionariesWithRecursionCheck(TCollection& result, TCollection& visited, bool recurse);
-   void                      GetMissingDictionariesForPairElements(TCollection& result, TCollection& visited, bool recurse);
+   TVirtualStreamerInfo *FindStreamerInfo(TObjArray* arr, UInt_t checksum) const;
```
Comment (pcanal) [inline core/meta/inc/TClass.h]: Please keep the original layout (function name aligned); clang-format is currently incapable to do what we really want in this area.

File: core/meta/inc/TClass.h [hunk 5]
```diff
@@ -309,12 +311,14 @@ friend class TProtoClass;
    TClass(const TClass& tc) = delete;
    TClass& operator=(const TClass&) = delete;
 
+   void LoadClassInfo() const;
```
Comment (pcanal) [inline core/meta/inc/TClass.h]: Why is moving LoadClassInfo from one part of the private section to another useful? (or am I missing something?)

File: core/meta/src/TClass.cxx [hunk 6]
```diff
@@ -1438,7 +1441,10 @@ void TClass::Init(const char *name, Version_t cversion,
          }
       }
       if (!fHasRootPcmInfo && gInterpreter->CheckClassInfo(fName, /* autoload = */ kTRUE)) {
-         gInterpreter->SetClassInfo(this);   // sets fClassInfo pointer
+         {
+            R__LOCKGUARD(gInterpreterMutex);
```
Comment (pcanal) [inline core/meta/src/TClass.cxx]: Is this needed here since it is also the very first statement of SetClassInfo?

File: core/meta/src/TClass.cxx [hunk 7]
```diff
@@ -3451,38 +3457,31 @@ const char *TClass::GetSharedLibs()
...
-            R__LOCKGUARD(gInterpreterMutex);
-            // NOTE: Add test to prevent redo if another thread has already done the work.
-            // if (!fHasRootPcmInfo) {
+   if (fState == kHasTClassInit) {
+      R__LOCKGUARD(gInterpreterMutex);
 
-            // The bases are in our ProtoClass; we don't need the class info.
-            TProtoClass *proto = TClassTable::GetProtoNorm(GetName());
-            if (proto && proto->FillTClass(this)) {
-               // Not sure this code is still needed
```
Comment (pcanal) [inline core/meta/src/TClass.cxx]: You removed this note, are you sure it is not needed?  I.e. is there a actual use case where this branch is taken (obviously I was not sure and needed some more time to investigate)

File: core/meta/src/TClass.cxx [hunk 8]
```diff
@@ -4448,46 +4448,49 @@ TVirtualStreamerInfo* TClass::GetStreamerInfo(Int_t version /* = 0 */) const
...
    newname += "@@emulated";
 
-   TClass *emulated = TClass::GetClass(newname);
+   R__LOCKGUARD(gInterpreterMutex);
 
-   TVirtualStreamerInfo* sinfo = 0;
+   TClass *emulated = TClass::GetClass(newname);
 
-   if (emulated) {
+   if (emulated)
```
Comment (pcanal) [inline core/meta/src/TClass.cxx]: Does clang-format really 'force' one line if statement to be without curly brackets?

File: core/meta/src/TClass.cxx [hunk 9]
```diff
@@ -4448,46 +4448,49 @@ TVirtualStreamerInfo* TClass::GetStreamerInfo(Int_t version /* = 0 */) const
...
 {
-   R__LOCKGUARD(gInterpreterMutex);
+   TVirtualStreamerInfo* sinfo = nullptr;
 
-   TString newname( GetName() );
+   TString newname(GetName());
    newname += "@@emulated";
 
-   TClass *emulated = TClass::GetClass(newname);
+   R__LOCKGUARD(gInterpreterMutex);
```
Comment (pcanal) [inline core/meta/src/TClass.cxx]: Nice catch (construct string earlier)

File: core/meta/src/TClass.cxx [hunk 10]
```diff
@@ -3451,38 +3451,31 @@ const char *TClass::GetSharedLibs()
 
 TList *TClass::GetListOfBases()
 {
-   if (!fBase) {
-      if (fCanLoadClassInfo) {
```
Comment (pcanal) [inline core/meta/src/TClass.cxx]: This 'if statement' was removed.  Could you add comments (at the very least in the commit) why it is not needed.

Comment (pcanal) [inline core/meta/src/TClass.cxx]: Also the commit mentioned a reduction in scope of the lock, as far as I can tell it is actually the same what am I missing?

File: core/meta/src/TClass.cxx [hunk 11]
```diff
@@ -3451,38 +3451,31 @@ const char *TClass::GetSharedLibs()
...
-         if (fState == kHasTClassInit) {
+   if (fBase)
```
Comment (pcanal) [inline core/meta/src/TClass.cxx]: This is indeed clearer and possibly faster.

File: core/meta/src/TClass.cxx [hunk 12]
```diff
@@ -1399,10 +1399,12 @@ void TClass::Init(const char *name, Version_t cversion,
    }
 
    if (givenInfo) {
-      if (!gInterpreter->ClassInfo_IsValid(givenInfo) ||
-          !(gInterpreter->ClassInfo_Property(givenInfo) & (kIsClass | kIsStruct | kIsNamespace)) ||
-          (!gInterpreter->ClassInfo_IsLoaded(givenInfo) && (gInterpreter->ClassInfo_Property(givenInfo) & (kIsNamespace))) )
-      {
+      R__LOCKGUARD(gInterpreterMutex);
```
Comment (pcanal) [inline core/meta/src/TClass.cxx]: This is clearer .. albeit in practice this is a no-op as Init is always called with the lock already taken (Init should be called only from the constructor and they should all be taking the lock before hand).

File: tree/tree/src/TTree.cxx [hunk 13]
```diff
@@ -4383,53 +4383,56 @@ Int_t TTree::Fill()
...
    Int_t nerror = 0;
-   Int_t nb = fBranches.GetEntriesFast();
-   if (nb == 1) {
-      // Case of one single super branch. Automatically update
-      // all the branch addresses if a new object was created.
-      TBranch* branch = (TBranch*) fBranches.UncheckedAt(0);
-      branch->UpdateAddress();
-   }
-   if (fBranchRef) {
+   Int_t nbranches = fBranches.GetEntriesFast();
```
Comment (pcanal) [inline tree/tree/src/TTree.cxx]: Most changes in the commit "Optimize TTree::Fill()" seems unrelated to the 'modulo operations' making it harder to understand ... is it possible to either split it or at least explain all the changes?

File: tree/tree/src/TTree.cxx [hunk 14]
```diff
@@ -4383,53 +4383,56 @@ Int_t TTree::Fill()
...
-         }
-         ++nerror;
-      } else {
+#endif
+
+      if (nwrite > 0) {
          nbytes += nwrite;
+      } else {
+         ++nerror;
+         Error("Fill", "Failed filling branch:%s.%s, nbytes=%d, entry=%lld\n"
```
Comment (pcanal) [inline tree/tree/src/TTree.cxx]: There is was intentionally code (if nerror<2) to reduce the output (by avoid repetition of a large piece of text).  What is the rational in removing it?

File: tree/tree/src/TTree.cxx [hunk 15]
```diff
@@ -4438,99 +4441,88 @@ Int_t TTree::Fill()
...
-               // Set fAutoSave to the largest integer multiple of
-               // fAutoFlush events such that fAutoSave*fFlushedBytes
-               // < (minus the input value of fAutoSave)
-               Long64_t totBytes = GetTotBytes();
-               if (zipBytes != 0) {
-                  fAutoSave =  TMath::Max( fAutoFlush, fEntries*((-fAutoSave/zipBytes)/fEntries));
-               } else if (totBytes != 0) {
-                  fAutoSave =  TMath::Max( fAutoFlush, fEntries*((-fAutoSave/totBytes)/fEntries));
-               } else {
-                  TBufferFile b(TBuffer::kWrite, 10000);
```
Comment (pcanal) [inline tree/tree/src/TTree.cxx]: This whole case seems to have disappeared (i.e. if GetToBytes()==0 then use a TBufferFile).  You must add a comment in the commit on what changed between the time it was written and now that explain why it is no longer needed.

File: core/meta/src/TClass.cxx [hunk 16]
```diff
@@ -4358,29 +4357,33 @@ Int_t TClass::GetNmethods()
...
+   // Version 0 is special, it means the currently loaded version.
+   // We need to set it at the beginning to be able to guess it correctly.
 
-   // Handle special version, 0 means currently loaded version.
-   // Warning:  This may be -1 for an emulated class.
-   // If version == -2, the user is requested the emulated streamerInfo
-   // for an abstract base class even though we have a dictionary for it.
-   if (version == 0) {
+   if (version == 0)
       version = fClassVersion;
```
Comment (pcanal) [inline core/meta/src/TClass.cxx]: Note fClassVersion is neither an atomic nor absolutely constant (See ROOT::ResetClassVersion) ... However it still may or may not be thread safe to use here dependenting on the usage pattern/requirement of ROOT::ResetClassVersion (and its callees or callers)

File: core/meta/src/TClass.cxx (hunk 2 above)
Comment (pcanal) [inline core/meta/src/TClass.cxx]: This looks fine but really hinge on whether the line number 4366 (in new file) is thread safe or not.

File: core/meta/inc/TClass.h (hunk 3 above)
Comment (amadio) [inline core/meta/inc/TClass.h]: Hi Philippe, thanks for reviewing my changes, even though you are still on vacations.

Although you are right, only one of those interpretations makes sense here, i.e., "the class represented by this instance of TClass". If it is for TClass itself, then TClass will be represented by this instance of TClass, and only "this class" does not disambiguate. In any case, I just got annoyed with the extra long lines and created a commit to bring line lengths down to 120 characters. I can drop this commit if it's deemed unnecessary/unwanted.

File: core/meta/inc/TClass.h (hunk 4 above)
Comment (amadio) [inline core/meta/inc/TClass.h]: The extra space is kinda ugly, and violates the coding conventions by making the lines too long. In any case, I will probably just drop the formatting commit and leave it for another time.

File: core/meta/inc/TClass.h (hunk 5 above)
Comment (amadio) [inline core/meta/inc/TClass.h]: You are right, I looked up in the file to check for public/private, but hit a `public:` label inside `InsertTClassInRegistryRAII` class that fooled me into thinking LoadClassInfo() was not private.
Good catch. I will drop this commit.

File: core/meta/src/TClass.cxx (hunk 6 above)
Comment (amadio) [inline core/meta/src/TClass.cxx]: Well, there was a lock around it in LoadClassInfo(), so when I saw it here I assumed we needed to lock too. Since there is already a lock there, we can remove it from both here and LoadClassInfo(), then, although it would be probably better to both auto parse and set class info all in one go by still taking the lock there.

File: core/meta/src/TClass.cxx (hunk 11 above)
Comment (amadio) [inline core/meta/src/TClass.cxx]: Yes, it's a lot faster. I edited this function because it appeared as a bottleneck in VTune.

File: core/meta/src/TClass.cxx (hunk 10 above)
Comment (amadio) [inline core/meta/src/TClass.cxx]: I added an early return, so the lock is taken much less often, but I agree that the commit message is not the best, so I will rewrite it and add a better explanation for the changes.

File: core/meta/src/TClass.cxx (hunk 7 above)
Comment (amadio) [inline core/meta/src/TClass.cxx]: No, we need to go over this more carefully. I want to discuss a way to add tests for all cases and when they are supposed to happen so that we can be sure in the future when changing this. Let me know what you think would be a good set of cases to start testing this function and I will add a commit with the test.

File: core/meta/src/TClass.cxx [hunk 17]
```diff
@@ -3451,38 +3457,31 @@ const char *TClass::GetSharedLibs()
...
+      if (proto && proto->FillTClass(this))
+         fHasRootPcmInfo = true;
+   }
 
-               fHasRootPcmInfo = kTRUE;
-            }
-         }
-         // We test again on fCanLoadClassInfo has another thread may have executed it.
-         if (!fHasRootPcmInfo && !fCanLoadClassInfo) {
-            LoadClassInfo();
```
Comment (amadio) [inline core/meta/src/TClass.cxx]: One thing I thought was strange was this checking on `!fCanLoadClassInfo` to call `LoadClassInfo()`. That's why I added `if (!fCanLoadClassInfo) return;` to `LoadClassInfo()` before. This check made no sense to me, also because it mentions other threads doing the work, in which case the function would have been done anyway, due to the lock being taken.

File: core/meta/src/TClass.cxx (hunk 16 above)
Comment (amadio) [inline core/meta/src/TClass.cxx]: If fClassInfo is not thread-safe, we need to make it thread-safe or go back to taking the lock, despite the performance penalty. At this point, seeing how many locks we have everywhere, I think we should seriously consider using RCU in ROOT instead of locks, as that is probably a much better solution in the long term.

File: core/meta/src/TClass.cxx (hunk 2 above)
Comment (amadio) [inline core/meta/src/TClass.cxx]: That line is apparently not thread-safe, due to false sharing (lack of alignment of fClassVersion), but real problems depend on if there is any pattern where modifications to the class version can actually happen when this function is called. As commented above, we may need to go back to taking the lock despite the performance penalty. On the other hand, we need a better long term solution for our thread-safety issues. On one side we need a test suite to catch thread-safety problems like these, and on the other, we need a better way than taking locks and using atomics.

File: core/meta/src/TClass.cxx (hunk 9 above)
Comment (amadio) [inline core/meta/src/TClass.cxx]: Thanks.

File: core/meta/src/TClass.cxx (hunk 8 above)
Comment (amadio) [inline core/meta/src/TClass.cxx]: No, I just find it cleaner to not have braces on if's with a single statement. Using less braces means that we can fit more code in a screenful of text, so it's easier to read the code too. However, `clang-format` does want to always put these if statements on a single line, which annoys me, as the indentation helps when reading.

File: core/meta/src/TClass.cxx [hunk 18]
```diff
@@ -5459,35 +5471,39 @@ TClass *TClass::LoadClassCustom(const char *requestedname, Bool_t silent)
...
 void TClass::LoadClassInfo() const
 {
-   R__LOCKGUARD(gInterpreterMutex);
+   bool autoParse = !gInterpreter->IsAutoParsingSuspended();
 
-   // If another thread executed LoadClassInfo at about the same time
-   // as this thread return early since the work was done.
-   if (!fCanLoadClassInfo) return;
+   {
+      R__LOCKGUARD(gInterpreterMutex);
```
Comment (amadio) [inline core/meta/src/TClass.cxx]: Both `AutoParse()` and `SetClassInfo()` below lock `gInterpreterMutex` themselves. Is there a point to taking the lock here for them to be coherent (i.e., avoid two waits for the lock), or should we simply remove the lock and let it be taken by the interpreter? It seems better to still take the lock, and actually extend it to the remaining of the function to ensure that no work is duplicated by other threads (that is, extend it until we set fCanLoadClassInfo to false, to trigger the early return in other threads).

File: tree/tree/src/TTree.cxx (hunk 13 above)
Comment (amadio) [inline tree/tree/src/TTree.cxx]: Sure, I will add explanations, as this is actually quite complicated to parse. There was a lot of branching in the previous code that I tried to reduce, since this is a very performance-critical function. New commit message coming, also with a link to the [JIRA issue](https://sft.its.cern.ch/jira/browse/ROOT-8872) where the analysis is being documented.

File: tree/tree/src/TTree.cxx (hunk 14 above)
Comment (amadio) [inline tree/tree/src/TTree.cxx]: Shouldn't errors be prominent in the output? What if the user is printing something and filling? They will likely miss the problem entirely.

File: tree/tree/src/TTree.cxx (hunk 15 above)
Comment (amadio) [inline tree/tree/src/TTree.cxx]: Indeed, I removed this on the basis that this part of the code is never reached when `fEntries == 0`, and if `fEntries > 0` either `GetZipBytes()` or `GetToBytes()` must return a positive number right after filling baskets, unless there is a serious bug in filling. Is my line of thought wrong? How can you actually have both `GetZipBytes()` and `GetToBytes()` return zero after flushing baskets?

File: core/meta/src/TClass.cxx (hunk 6 above)
Comment (pcanal) [inline core/meta/src/TClass.cxx]: > although it would be probably better to both auto parse and set class info all in one go by still taking the lock there.

Yes, this is what the already happens in LoadClassInfo.  I do not see how this lock does anything (since it is limited to a single statement (which turns around and takes the lock a second time) by the explicit scope).

File: core/meta/src/TClass.cxx [hunk 19]
```diff
@@ -5459,35 +5471,39 @@ TClass *TClass::LoadClassCustom(const char *requestedname, Bool_t silent)
...
-      fCanLoadClassInfo = kFALSE;
+   if (autoParse && !fClassInfo) {
+      ::Error("TClass::LoadClassInfo",
+            "no interpreter information for class %s is available"
+            " even though it has a TClass initialization routine.",
+            fName.Data());
+      return;
    }
+
+   fCanLoadClassInfo = false;
```
Comment (pcanal) [inline core/meta/src/TClass.cxx]: Doing this outside the lock has an 'interesting' side effect.  It allows two thread to do the work twice (once each).

if two thread get to LoadClassInfo at the same time, one will take the lock, do the autoparsing and the SetClassInfo, then release the lock .. and let's say get suspended.  The second thread then get the lock and since fCanLoadClassInfo has not been changed yet, will do the autoparsing again and the SetClassInfo again ...

File: core/meta/src/TClass.cxx (hunk 17 above)
Comment (pcanal) [inline core/meta/src/TClass.cxx]: > One thing I thought was strange was this checking on !fCanLoadClassInfo to call LoadClassInfo(). 

In the contrary, this is the intended pattern.  i.e. Call the function only if it might be useful.  All Calls to LoadClassInfo ought to be protected by that if statement.

File: core/meta/src/TClass.cxx (hunk 7 above)
Comment (pcanal) [inline core/meta/src/TClass.cxx]: short answer: create a series of test calling GetListOfBases with TClass in all possibles states (including missing dictionary, interpreted, emulated, with and without pcm, etc... and combination thereof)

File: core/meta/src/TClass.cxx (hunk 17 above)
Comment (pcanal) [inline core/meta/src/TClass.cxx]: > This check made no sense to me, also because it mentions other threads doing the work, 

The existing code is
```
if (fCanLoadClassInfo) {
   if (...) {
      R__LOCKGUARD(...);
   }
   if (!fHasRootPcmInfo && !fCanLoadClassInfo) {
```
The comment says that in-between the first line and the last line, another thread may have successfully run LoadClassInfo and thus, even-though we already tested for the value of fCanLoadClassInfo, we should test for it again.


On another note the question is why did you remove the test on '!fHasRootPcmInfo'

File: core/meta/src/TClass.cxx (hunk 16 above)
Comment (pcanal) [inline core/meta/src/TClass.cxx]: > If fClassInfo is not thread-safe
Here we are talking about fClassVersion :)
> At this point, seeing how many locks we have everywhere, I think we should seriously consider using RCU in ROOT instead of locks,
In general as far as I can tell this is indeed what we have been going towards where we can.  When more than one entity need to be updated 'at the same time' (eg. the interpreter states) we have no choice but to take lock ..

>  we need to make it thread-safe or go back to taking the lock, despite the performance penalty. 

Yes my point exactly :)

File: core/meta/src/TClass.cxx (hunk 18 above)
Comment (pcanal) [inline core/meta/src/TClass.cxx]: > Is there a point to taking the lock here for them to be coherent

yes, it is more coherent and avoid duplication of work.

> and actually extend it to the remaining of the function to ensure

Yes (and this is actually the current implementation).

File: core/meta/src/TClass.cxx (hunk 6 above)
Comment (amadio) [inline core/meta/src/TClass.cxx]: I agree. I didn't know before that SetClassInfo was taking the lock. The way to go here is to combine both sections taking the lock together, as done in `LoadClassInfo()`.

File: core/meta/src/TClass.cxx (hunk 19 above)
Comment (amadio) [inline core/meta/src/TClass.cxx]: Yes, I also noticed that yesterday and agree with you. The lock must be taken for the whole duration of the function, so that the next thread will exit upon taking the lock on the second check to `fCanLoadClassInfo == true`.

File: core/meta/src/TClass.cxx (hunk 17 above)
Comment (amadio) [inline core/meta/src/TClass.cxx]: The point is that if `fCanLoadClassInfo == false`, then calling the function *never* changes the result of `fClassInfo`, and the check was inverted (by checking for `!fCanLoadClassInfo` instead of `fCanLoadClassInfo`).

File: core/meta/src/TClass.cxx (hunk 7 above)
Comment (amadio) [inline core/meta/src/TClass.cxx]: Ok. These tests should already have been created, but better late than never :smile:

**Comment by amadio** [comment]
Alright, thanks for all the feedback! I think that it will be better to break this PR into smaller pieces, so I will close this one, rework the commits according to the feedback received, and push new PRs soon.
